src/copc.rs

```diff
@@ -0,0 +1,395 @@
+    pub const USER_ID: &'static str = "copc";
+    /// The record id of the LasZip VLR header.
+    pub const RECORD_ID: u16 = 1;
+    /// The description of the LasZip VLR header.
+    pub const DESCRIPTION: &'static str = "https://copc.io";
+
+    /// Reads the Vlr data from the source.
+    ///
+    /// This **only** reads the *payload data* the
+    /// vlr header should already be read.
+    fn read_from<R: Read>(mut src: R) -> Result<Self> {
+        Ok(Self {
+            center_x: src.read_f64::<LittleEndian>()?,
+            center_y: src.read_f64::<LittleEndian>()?,
+            center_z: src.read_f64::<LittleEndian>()?,
+            halfsize: src.read_f64::<LittleEndian>()?,
+            spacing: src.read_f64::<LittleEndian>()?,
+            root_hier_offset: src.read_u64::<LittleEndian>()?,
+            root_hier_size: src.read_u64::<LittleEndian>()?,
+            gpstime_minimum: src.read_f64::<LittleEndian>()?,
+            gpstime_maximum: src.read_f64::<LittleEndian>()?,
+            reserved: {
+                let mut reserved = [0; 11];
+                for field in reserved.iter_mut() {
+                    *field = src.read_u64::<LittleEndian>()?;
+                }
+                reserved
+            },
+        })
+    }
+
+    /// Writes the Vlr data to the source.
+    ///
+    /// This **only** writes the *payload data* the
+    /// vlr header should be written before-hand.
+    pub fn write_to<W: Write>(&self, dst: &mut W) -> Result<()> {
+        dst.write_f64::<LittleEndian>(self.center_x)?;
+        dst.write_f64::<LittleEndian>(self.center_y)?;
+        dst.write_f64::<LittleEndian>(self.center_z)?;
+        dst.write_f64::<LittleEndian>(self.halfsize)?;
+        dst.write_f64::<LittleEndian>(self.spacing)?;
+        dst.write_u64::<LittleEndian>(self.root_hier_offset)?;
+        dst.write_u64::<LittleEndian>(self.root_hier_size)?;
+        dst.write_f64::<LittleEndian>(self.gpstime_minimum)?;
+        dst.write_f64::<LittleEndian>(self.gpstime_maximum)?;
+        self.reserved
+            .into_iter()
+            .try_for_each(|i| dst.write_u64::<LittleEndian>(i))?;
+        Ok(())
+    }
+}
+
+impl TryFrom<&Vlr> for CopcInfoVlr {
+    type Error = Error;
+
+    fn try_from(value: &Vlr) -> Result<Self> {
+        Self::read_from::<&[u8]>(value.data.as_ref())
+    }
+}
+
+/// VoxelKey corresponds to the naming of EPT data files.
+/// <https://entwine.io/en/latest/entwine-point-tile.html#ept-data>
+/// The point cloud data itself is arranged in a 3D analogous manner to slippy map tiling schemes.
+/// The scheme is Level-X-Y-Z.
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct VoxelKey {
+    // A value < 0 indicates an invalid VoxelKey
+    l: i32,
+    x: i32,
+    y: i32,
+    z: i32,
+}
+impl VoxelKey {
+    /// Computes the Childs of a VoxelKey
+    /// There are max 8 Childs to a VoxelKey
+    /// **dir**ection needs to be 0..8
+    pub fn child(&self, dir: i32) -> Self {
```

Comment:
Letters are cheap, let's avoid abbreviations if we can.

```suggestion
    /// **dir**ection needs to be 0..8
    pub fn child(&self, direction: i32) -> Self {
```